x/ccv/provider/keeper/key_assignment.go

```diff
@@ -0,0 +1,618 @@
+}
+
+// AppendConsumerAddrsToPrune appends a consumer validator address to the list of consumer addresses
+// that can be pruned once the VSCMaturedPacket with vscID is received
+func (k Keeper) AppendConsumerAddrsToPrune(ctx sdk.Context, chainID string, vscID uint64, consumerAddr sdk.ConsAddress) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.ConsumerAddrsToPruneKey(chainID, vscID))
+	var consumerAddrsToPrune types.AddressList
+	if bz != nil {
+		err := consumerAddrsToPrune.Unmarshal(bz)
+		if err != nil {
+			panic(err)
+		}
+	}
+	consumerAddrsToPrune.Addresses = append(consumerAddrsToPrune.Addresses, consumerAddr)
+	bz, err := consumerAddrsToPrune.Marshal()
+	if err != nil {
+		panic(err)
+	}
+	store.Set(types.ConsumerAddrsToPruneKey(chainID, vscID), bz)
+}
+
+// GetConsumerAddrsToPrune returns the list of consumer addresses
+// that can be pruned once the VSCMaturedPacket with vscID is received
+func (k Keeper) GetConsumerAddrsToPrune(ctx sdk.Context, chainID string, vscID uint64) [][]byte {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.ConsumerAddrsToPruneKey(chainID, vscID))
+	if bz == nil {
+		return nil
+	}
+	var consumerAddrsToPrune types.AddressList
+	err := consumerAddrsToPrune.Unmarshal(bz)
+	if err != nil {
+		panic(err)
+	}
+	return consumerAddrsToPrune.Addresses
+}
+
+func (k Keeper) IterateConsumerAddrsToPrune(
+	ctx sdk.Context,
+	chainID string,
+	cb func(vscID uint64, consumerAddrsToPrune [][]byte) (stop bool),
+) {
+	store := ctx.KVStore(k.storeKey)
+	iteratorPrefix := types.ChainIdWithLenKey(types.ConsumerAddrsToPruneBytePrefix, chainID)
+	iterator := sdk.KVStorePrefixIterator(store, iteratorPrefix)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		_, vscID, err := types.ParseChainIdAndVscIdKey(types.ConsumerAddrsToPruneBytePrefix, iterator.Key())
+		if err != nil {
+			panic(err)
+		}
+		var consumerAddrsToPrune types.AddressList
+		err = consumerAddrsToPrune.Unmarshal(iterator.Value())
+		if err != nil {
+			panic(err)
+		}
+		stop := cb(vscID, consumerAddrsToPrune.Addresses)
+		if stop {
+			break
+		}
+	}
+}
+
+func (k Keeper) IterateAllConsumerAddrsToPrune(
+	ctx sdk.Context,
+	cb func(chainID string, vscID uint64, consumerAddrsToPrune [][]byte) (stop bool),
+) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte{types.ConsumerAddrsToPruneBytePrefix})
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		chainID, vscID, err := types.ParseChainIdAndVscIdKey(types.ConsumerAddrsToPruneBytePrefix, iterator.Key())
+		if err != nil {
+			panic(err)
+		}
+		var consumerAddrsToPrune types.AddressList
+		err = consumerAddrsToPrune.Unmarshal(iterator.Value())
+		if err != nil {
+			panic(err)
+		}
+		stop := cb(chainID, vscID, consumerAddrsToPrune.Addresses)
+		if stop {
+			break
+		}
+	}
+}
+
+// DeleteConsumerAddrsToPrune deletes the list of consumer addresses mapped to a given VSC ID
```

Comment:
In this docstring, or somewhere else, the pruning invariant should be written. In other terms, when _exactly_ is it possible to get a provider addr from a given consumer addr using `GetValidatorByConsumerAddr` . The test can then be phrased similarly to how it was before

https://github.com/cosmos/interchain-security/blob/4743b4dbb7765647692574d3320b07806b51060c/x/ccv/provider/keeper/key_assignment_test.go#L329-L348

Reply:
What about

'
For all consumer addr CK in ValidatorByConsumerAddr:
  either (
  (
    there exists a provider addr PK in ValidatorConsumerPubKey st. ValidatorConsumerPubKey(PK) = CK
   ) or
  (
   there exists a vscid VID in ConsumerAddrsToPrune st. CK in ConsumerAddrsToPrune(VID)
  )
)
'

Reply:
I added it to `AppendConsumerAddrsToPrune`